ssm_jax/bp/gauss_bp.py

```diff
@@ -0,0 +1,212 @@
+from functools import partial
+from jax import jit
+from jax import numpy as jnp
+from jax.tree_util import tree_map
+
+
+def block_split(A, idx):
+    """Split square matrix A into four blocks at `idx`
+
+    For example splitting
+        [[0, 1, 2],
+         [3, 4, 5],
+         [6, 7 ,8]]
+    at idx = 2 produces:
+        ([[0,1],[3,4]],
+         [[2],[5]],
+         [[6,7]],
+         [[8]]).
+
+    Args:
+        A (D, D): array.
+        idx: int.
+    Returns:
+        A11 (dim1, dim1)
+        A12 (dim2, dim1)
+        A21 (dim1, dim2)
+        A22 (dim2, dim2)
+       where `dim1 + dim2 == D` and `dim2 = D - idx`.
+    """
+    split_array = jnp.array([idx])
+    vblocks = jnp.vsplit(A, split_array)
+    # [leaf for tree in forest for leaf in tree]
+    blocks = [block for vblock in vblocks for block in jnp.hsplit(vblock, split_array)]
+    # Can also do:
+    #   blocks = tree_map(lambda arr: jnp.hsplit(arr,split_array),vblocks)
+    # followed by a tree_flatten
+    return blocks
+
+
+def block_join(A11, A12, A21, A22):
+    return jnp.block([[A11, A12], [A21, A22]])
+
+
+def block_rev(A, idx):
+    blocks = block_split(A, idx)
+    return block_join(*blocks[::-1])
+
+
+def vec_swap(a, idx):
+    return jnp.concatenate((a[idx:], a[:idx]))
+
+
+def info_marginalise(K_blocks, hs):
```

Comment:
Maybe it would be clearer to pass in `Kxx, Kxy, Kyy, hx, hy` explicity, instead of a tuple?